[Kic2.0] Fix KIC image deployment failure #1486

ccfishk opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md and ensure your changes are being reflected in CHANGELOG.md for the next upcoming release
-->

**What this PR does / why we need it**:

Able to deploy kong ingress controller into the cluster namespace using a yaml and controller is able to run and function as expected.



**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

#1480 
#1421 
#1256 


**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR
- [] able to run performance/integration tests with incluster deployment
- [] all manifests and related service accounts have proper RBAC roles associated, none of them use ClusterAdmin permissions

rainest commented:
Do we know where the attempt to edit `controller-token-tdslg` is coming from? I haven't encountered that before and am not sure where/why the controller would need to edit a Secret. I'm not clear on the link between it and the changes proposed here.

Historically we haven't had any system to automatically infer the admin API URL from a Service or similar, and hadn't had any plans to add one, as directly providing a URL has been sufficient. Can you elaborate on why you added it? The proposal here makes several decisions we wouldn't want (always using HTTP and sending traffic through an Ingress rather than to a Service over the intra-cluster network) that the URL avoids.

Manifests we planned to handle in #1256, though that's more about building out new kustomize helpers than just providing a manifest--the existing 1.x manifests should work fine for now.

shaneutt commented on railgun/deployment/kic-deployment.template.yaml:
```diff
@@ -0,0 +1,1095 @@
+          periodSeconds: 10
+          successThreshold: 1
+          timeoutSeconds: 1
+      serviceAccountName: kong-serviceaccount
```
blocker: please _do not include this new deployment mechanism in this PR_, https://github.com/Kong/kubernetes-ingress-controller/issues/1480 is a `POST KIC 2.0` item that we need to discuss as a team. The team needs to have consensus that we want to support yet another deployment mechanism, and we shouldn't be adding any scope to KIC 2.0 that doesn't have ties to a direct product/customer request or issue.

ccfishk replied:
This is not a new mechanism at all. You can take it as parity - user deploy v1.3.x kong (both controller and proxy) using one yaml https://bit.ly/k4k8s.

In 2.0, we can either deploy controller along with proxy or individually. From long term, the benefits of individual yaml is pretty straightforward 
- aligned split kong-proxy/admin into kong-system
- less dependency/loose couply, if proxy fails to be up, controller does not get impacted
- scalable, resources /cpu/memory/db/network more configurable

btw, yaml deployment is generic k8s deployment.

shaneutt replied:
Gotcha, thanks for that explanation. That functionality (the `all-in-one` yaml) is actually just using our Kustomize build:

https://github.com/Kong/kubernetes-ingress-controller/blob/main/hack/build-single-manifests.sh

If the intention is to recreate that functionality then I feel like we should update that script with our new kustomize files, rather than creating a new manifest we have to maintain and update manually. How does that sound to you what are your thoughts?

shaneutt commented on railgun/internal/ctrlutils/utils.go:
```diff
@@ -141,3 +146,31 @@ func CRDExists(client client.Client, gvr schema.GroupVersionResource) bool {
 	return true
 }
+
+// retrieve Kong Admin API URL from configured name/namespace service
+func RetrieveKongAdminAPIURL(ctx context.Context, KongAdminAPI string, kubeCfg *rest.Config) (string, error) {
```
`godoc` comments for a symbol must begin with the symbol they are documenting:
```suggestion
// RetrieveKongAdminAPIURL retrieves the Kong Admin API URL from configured name/namespace service
func RetrieveKongAdminAPIURL(ctx context.Context, KongAdminAPI string, kubeCfg *rest.Config) (string, error) {
```

ccfishk replied:
Good catch

shaneutt commented on railgun/internal/ctrlutils/utils.go:
```diff
@@ -141,3 +146,31 @@ func CRDExists(client client.Client, gvr schema.GroupVersionResource) bool {
+func RetrieveKongAdminAPIURL(ctx context.Context, KongAdminAPI string, kubeCfg *rest.Config) (string, error) {
+	namespace, name, err := util.ParseNameNS(KongAdminAPI)
+	if err != nil {
+		return "", fmt.Errorf("failed to pars kong admin api namespace and name. err %v", err)
```
There's a typo present, and we can use `%w` to created a [wrapped error](https://golang.org/pkg/fmt/#Errorf) which is more consistent with the rest of the codebase:
```suggestion
		return "", fmt.Errorf("failed to parse kong admin api namespace and name: %w", err)
```

shaneutt commented on railgun/internal/ctrlutils/utils.go:
```diff
@@ -141,3 +146,31 @@ func CRDExists(client client.Client, gvr schema.GroupVersionResource) bool {
+	if err != nil {
+		return "", fmt.Errorf("failed to pars kong admin api namespace and name. err %v", err)
+	}
+	CoreClient, _ := clientset.NewForConfig(kubeCfg)
```
There are two problems we need to solve here:

1. we're throwing away the error, which causes fragility which could result in runtime `panics`
2. we don't need to create a new client each time this message is called if the `*kube.Config` is static, we can simply send in the client itself to reduce calls and memory utilization.

My suggestion is to employ number 2 above, and make sure that we check the error.

ccfishk replied:
agree that we could not swallow error silently.

shaneutt commented on railgun/internal/ctrlutils/utils.go:
```diff
@@ -141,3 +146,31 @@ func CRDExists(client client.Client, gvr schema.GroupVersionResource) bool {
+	CoreClient, _ := clientset.NewForConfig(kubeCfg)
+	svc, err := CoreClient.CoreV1().Services(namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to creating kubernetes client")
```
This error message is not accurate, because the error is coming from the `Services().Get`, not the clienset creation. Also we can wrap the encountered error for more logging visibility:
```suggestion
		return "", fmt.Errorf("failed retrieve service object %s/%s: %w", namespace, name, err)
```

shaneutt commented on railgun/internal/ctrlutils/utils.go:
```diff
@@ -141,3 +146,31 @@ func CRDExists(client client.Client, gvr schema.GroupVersionResource) bool {
+		return "", fmt.Errorf("failed to creating kubernetes client")
+	}
+	ingresses := svc.Status.LoadBalancer.Ingress
+	adminIP := ""
+	for _, ingress := range ingresses {
+		adminIP = ingress.IP
+	}
```
This implementation is bound to work most if not all the time, however it has two technical flaws:

1. it can technically _leave the `adminIP` as an empty string `""`_
2. it doesn't account for the technical possibility of multiple `Ingress` entries

My recommendation would be to validate that the list is exactly 1 in length, and throw an error if not.

ccfishk replied:
A loadbalancer type can not have multiple external ip/host, otherwise it is against ip uniqueness.

shaneutt replied:
setting aside multiple IPs, it could still be an empty list right? So that could be a problem we could do a short validation check for?

shaneutt commented on railgun/internal/ctrlutils/utils.go:
```diff
@@ -141,3 +146,31 @@ func CRDExists(client client.Client, gvr schema.GroupVersionResource) bool {
+		adminIP = ingress.IP
+	}
+
+	ports := svc.Spec.Ports
+	var adminPort int32
+	for _, port := range ports {
+		if port.Name == "kong-admin" {
+			adminPort = port.Port
+		}
+	}
```
This is another place some light validation could improve the functionality of this code, I suggest we validate the `Spec.Ports` here to make sure it's in order and throw an error if not, so we don't end up with adminPort being set to the default int value of 0, which could break things in other places.

ccfishk replied:
I dont believe user is able to set a k8s resource port( targetport) as 0, that is against k8s rule.

shaneutt replied:
Correct, the _user_ could not. But in this case, it's the _code_ which can try to set it to `0` right? If we don't validate there's something there, the value will be initialized as 0 as this is the default for the `int32` var `adminPort`?

shaneutt commented on railgun/internal/ctrlutils/utils.go:
```diff
@@ -141,3 +146,31 @@ func CRDExists(client client.Client, gvr schema.GroupVersionResource) bool {
+			adminPort = port.Port
+		}
+	}
+	kongAdminURL := fmt.Sprintf("http://%s:%d", adminIP, adminPort)
+	return kongAdminURL, nil
```
nit: you can return the result of `fmt.Sprint` rather than store it if there's nothing else that's going to use the variable:
```suggestion
	return fmt.Sprintf("http://%s:%d", adminIP, adminPort), nil
```

ccfishk replied:
Good catch.

shaneutt commented on railgun/pkg/config/config.go:
```diff
@@ -55,6 +57,9 @@ type Config struct {
 	FilterTags           []string
 	WatchNamespace       string
 
+	PublishService string
+	KongAdminAPI   string
```
nit: elsewhere in this code we have comments that at least group together related configurations. Please add a comment here that provides some general explanation of why these flags exist and why they're grouped together.

shaneutt commented on railgun/pkg/config/config.go:
```diff
@@ -197,3 +207,42 @@ func (c *Config) GetKubeClient() (client.Client, error) {
 	}
 	return client.New(conf, client.Options{})
 }
+
+func (c *Config) ConfigKongService(ctx context.Context) error {
```
lint: public symbols need godoc comments

shaneutt commented on railgun/pkg/config/config.go:
```diff
@@ -197,3 +207,42 @@ func (c *Config) GetKubeClient() (client.Client, error) {
+		if len(c.KongAdminURL) > 0 && c.KongAdminURL != "http://localhost:8001" {
+			return nil
+		}
+		return fmt.Errorf(" kong admin api url %s should not be localhost.", c.KongAdminURL)
```
question: what led us to make this assertion that the kong admin API URL should not be localhost? It seems to me that we should be allowing this, and making this change I worry will in fact break things? :thinking:

rainest replied:
The function is a bit confusing, since it manipulates the user config struct (which we want to get away from) as a side effect. I think the error text is inaccurate--it looks like the effect here is instead to _only_ allow `c.KongAdminURL == "http://localhost:8001"` if it is set along with `CONTROLLER_KONG_ADMIN_PUBLISH_SERVICE`. If so, `ConfigKongService()` exits _without_ modifying the config struct, i.e. if `c.KongAdminURL` was set to `http://localhost:8001`, it remains so. If those conditions aren't true, `ConfigKongService()` proceeds, composes a URL from the Service, and updates the config struct with that URL.

I don't think this behavior is correct (it doesn't handle common cases like a non-standard admin API port or HTTPS admin API listen on localhost, and relies on extracting configuration directly from the environment instead of using the `config` package--something we should only use for test environment special cases), but can't really evaluate that without a clearer understanding of the design goals here, hence my [original question](https://github.com/Kong/kubernetes-ingress-controller/pull/1487#issuecomment-874862259) asking why we'd add this in the first place.

ccfishk replied:
Here is the clarification: 
assuming admin api url configuration "http://localhost:8001" is valid and generic configuration, if deployments does not specify configuration (admin api url service name and namespace), leave as it is line 216. 

Otherwise, specified name/namespace will be applied to retrieve the exact service IP/hostname and port, which is used to compose the in cluster URL i.e  http://172.18.0.240:8001

The whole point is to read information from in cluster configuration (no hardcode).

shaneutt commented on railgun/pkg/config/config.go:
```diff
@@ -197,3 +207,42 @@ func (c *Config) GetKubeClient() (client.Client, error) {
+
+	kubeCfg, err := c.GetKubeconfig()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve kubeconfig. err %v", err)
```
to be more consistent with the rest of the code, we should do error wrapping: 
```suggestion
		return fmt.Errorf("failed to retrieve kubeconfig: %w", err)
```

shaneutt commented on railgun/pkg/config/config.go:
```diff
@@ -197,3 +207,42 @@ func (c *Config) GetKubeClient() (client.Client, error) {
+	}
+	kongadminurl, err := ctrlutils.RetrieveKongAdminAPIURL(ctx, adminApiService, kubeCfg)
+	if err != nil || len(kongadminurl) == 0 {
+		return fmt.Errorf("failed to generating kong admin url. err %v", err)
```
to be more consistent with the rest of the code, we should do error wrapping:
```suggestion
		return fmt.Errorf("failed to generating kong admin url: %w", err)
```

shaneutt commented on railgun/pkg/config/config.go:
```diff
@@ -197,3 +207,42 @@ func (c *Config) GetKubeClient() (client.Client, error) {
+	}
+	c.KongAdminURL = kongadminurl
+
+	if len(c.PublishService) == 0 {
+		kongService := os.Getenv("CONTROLLER_PUBLISH_SERVICE")
```
question: are we setting a _new precedent_ here that the flags will override ENV vars?

Would it potentially be better for us to pre-emptively parse ENV vars and configuring the *config.Config option accordingly in a single place that's meant for this purpose? :thinking:

ccfishk replied:
We don't overwrite env vars. On the other way, we read env vars and use it configure the kong Config options.

ccfishk commented:
> #1256



> Do we know where the attempt to edit `controller-token-tdslg` is coming from? I haven't encountered that before and am not sure where/why the controller would need to edit a Secret. I'm not clear on the link between it and the changes proposed here.
> 
> Historically we haven't had any system to automatically infer the admin API URL from a Service or similar, and hadn't had any plans to add one, as directly providing a URL has been sufficient. Can you elaborate on why you added it? The proposal here makes several decisions we wouldn't want (always using HTTP and sending traffic through an Ingress rather than to a Service over the intra-cluster network) that the URL avoids.
> 
> Manifests we planned to handle in #1256, though that's more about building out new kustomize helpers than just providing a manifest--the existing 1.x manifests should work fine for now.

Please ignore the specific error messages (sorry for that example details) since different ones pop up when debugging deployment. The purpose of this PR is to unblock controller manager in cluster deployment with appropriate configuration which is illustrated in the yaml and readme file. 

The points of the Kong Admin API URL is to remove hardcode ip address (publishStatusAddress) and leverage environment variable retrieving service information. Understand the previous configuration due to controller and proxy are in the same namespace and same pods. Now, we have the flexibility to de-couple them in 2.0

rainest commented:
> unblock controller manager in cluster deployment with appropriate configuration which is illustrated in the yaml and readme file.

I was trying to understand what issue we had there though--if there is some error that appears, we'd want to understand why it occurs to ensure we're adding a fix. I wasn't aware of any, as when I test the controller 2.0.0-alpha.1 image with the current chart and the all-in-one manifests we have in https://github.com/Kong/kubernetes-ingress-controller/tree/main/deploy/single it does deploy successfully. I am aware of issues with the RBAC config and lack of UDPIngress in the existing examples, but I don't think they cause failures (we just can't update finalizers and have to disable UDPIngress), and we planned to update that elsewhere (#1256 and https://github.com/Kong/charts/pull/364). The rest of the manifests that we had for 1.x do still work with 2.x.

> The points of the Kong Admin API URL is to remove hardcode ip address (publishStatusAddress) and leverage environment variable retrieving service information. Understand the previous configuration due to controller and proxy are in the same namespace and same pods.

This sounds related to the work in https://github.com/Kong/kubernetes-ingress-controller/pull/1451, correct? fe24a710b41ea07d48688f9290706309d8762ea3 should have already given us what we need there. `publishStatusAddress` should remain: it serves as a user override for Ingress status addresses when they know there network topology doesn't let the controller effectively determine the correct address information (similar to the [Service clusterIP setting](https://kubernetes.io/docs/concepts/services-networking/service/#choosing-your-own-ip-address)), as an alternative to `--publish-service`.

That doesn't have any interaction with the admin API, as we're only concerned with the proxy for Ingress status. We shouldn't need any additional admin API logic--while building configuration from a Service instead is viable, it doesn't really give us anything obvious over requesting a URL instead (since we don't need to look up endpoints or obtain other detailed information about the Service), and our documentation all already refers to the existing `--kong-admin-url` flag.

mflendrich commented:
Just a note: the manifest template probably overlaps with https://github.com/Kong/kubernetes-ingress-controller/pull/1500

ccfishk commented:
> > unblock controller manager in cluster deployment with appropriate configuration which is illustrated in the yaml and readme file.
> 
> I was trying to understand what issue we had there though--if there is some error that appears, we'd want to understand why it occurs to ensure we're adding a fix. I wasn't aware of any, as when I test the controller 2.0.0-alpha.1 image with the current chart and the all-in-one manifests we have in https://github.com/Kong/kubernetes-ingress-controller/tree/main/deploy/single it does deploy successfully. I am aware of issues with the RBAC config and lack of UDPIngress in the existing examples, but I don't think they cause failures (we just can't update finalizers and have to disable UDPIngress), and we planned to update that elsewhere (#1256 and [Kong/charts#364](https://github.com/Kong/charts/pull/364)). The rest of the manifests that we had for 1.x do still work with 2.x.
> 
> > The points of the Kong Admin API URL is to remove hardcode ip address (publishStatusAddress) and leverage environment variable retrieving service information. Understand the previous configuration due to controller and proxy are in the same namespace and same pods.
> 
> This sounds related to the work in #1451, correct? [fe24a71](https://github.com/Kong/kubernetes-ingress-controller/commit/fe24a710b41ea07d48688f9290706309d8762ea3) should have already given us what we need there. `publishStatusAddress` should remain: it serves as a user override for Ingress status addresses when they know there network topology doesn't let the controller effectively determine the correct address information (similar to the [Service clusterIP setting](https://kubernetes.io/docs/concepts/services-networking/service/#choosing-your-own-ip-address)), as an alternative to `--publish-service`.
> 
> That doesn't have any interaction with the admin API, as we're only concerned with the proxy for Ingress status. We shouldn't need any additional admin API logic--while building configuration from a Service instead is viable, it doesn't really give us anything obvious over requesting a URL instead (since we don't need to look up endpoints or obtain other detailed information about the Service), and our documentation all already refers to the existing `--kong-admin-url` flag.



> > unblock controller manager in cluster deployment with appropriate configuration which is illustrated in the yaml and readme file.
> 
> I was trying to understand what issue we had there though--if there is some error that appears, we'd want to understand why it occurs to ensure we're adding a fix. I wasn't aware of any, as when I test the controller 2.0.0-alpha.1 image with the current chart and the all-in-one manifests we have in https://github.com/Kong/kubernetes-ingress-controller/tree/main/deploy/single it does deploy successfully. I am aware of issues with the RBAC config and lack of UDPIngress in the existing examples, but I don't think they cause failures (we just can't update finalizers and have to disable UDPIngress), and we planned to update that elsewhere (#1256 and [Kong/charts#364](https://github.com/Kong/charts/pull/364)). The rest of the manifests that we had for 1.x do still work with 2.x.
> 
> > The points of the Kong Admin API URL is to remove hardcode ip address (publishStatusAddress) and leverage environment variable retrieving service information. Understand the previous configuration due to controller and proxy are in the same namespace and same pods.
> 
> This sounds related to the work in #1451, correct? [fe24a71](https://github.com/Kong/kubernetes-ingress-controller/commit/fe24a710b41ea07d48688f9290706309d8762ea3) should have already given us what we need there. `publishStatusAddress` should remain: it serves as a user override for Ingress status addresses when they know there network topology doesn't let the controller effectively determine the correct address information (similar to the [Service clusterIP setting](https://kubernetes.io/docs/concepts/services-networking/service/#choosing-your-own-ip-address)), as an alternative to `--publish-service`.
> 
> That doesn't have any interaction with the admin API, as we're only concerned with the proxy for Ingress status. We shouldn't need any additional admin API logic--while building configuration from a Service instead is viable, it doesn't really give us anything obvious over requesting a URL instead (since we don't need to look up endpoints or obtain other detailed information about the Service), and our documentation all already refers to the existing `--kong-admin-url` flag.

Agree that both publish-service and PublicStatusAddress should retain as it is, which was covered by https://github.com/Kong/kubernetes-ingress-controller/pull/1509 and https://github.com/Kong/kubernetes-ingress-controller/pull/1451. It is unrelated to the KIC individual deployment yaml

KIC does need KongAdminAPI for configuration communication, 
https://github.com/Kong/kubernetes-ingress-controller/blob/main/railgun/manager/run.go#L96

Current configuration way https://github.com/Kong/kubernetes-ingress-controller/pull/1500/files#L90 assumes all modules are in the same node and having static port (8444), this will break if we deploy controller and proxy separately. The function retrieve Admin ip and port through namespace and name configuration. https://github.com/Kong/kubernetes-ingress-controller/pull/1487/files#diff-786a1651aa544e8c2f84a32b474758ecb0f8e6c5d9d745b5f643d3d786bc57c4R152

rainest commented:
The assumptions in the various all-in-one YAMLs are by design: those are intentionally opinionated configurations for a select few common configurations (OSS or Enterprise, either with or without a database). They use a single Pod with one Kong container and a controller container because that topology is simple to manage, suitable for many users (especially those that don't have a known reason to use some other topology), and has security benefits for OSS (placing the controller in the same Pod means you don't need to expose the admin API, which is a simple means of securing it without RBAC).

If you want another topology, we currently instruct you to use the chart or to build your own manifests by hand. That stems from a combination of historical reasons (when we started providing manifests, Helm was fairly mature and both Kustomize and Operators were new) and capacity (we have limited resources to maintain deployment tools, so have not built out more complex Kustomize manifests or a non-Helm Operator). We don't provide plain manifests for topologies other than single-Pod because there are so many of them (you can basically mix and match any of the ~12 options covered in the [chart documentation](https://github.com/Kong/charts/blob/main/charts/kong/README.md#deployment-options), and even the chart doesn't commit to handling all topologies in their entirety--many require multiple releases).

That all said, I still don't think we gain anything by building admin API information from a Service rather than providing a URL, and introduces new problems (which aren't solved here yet):

* Using a Service name/namespace doesn't remove the requirement to indicate where the Service resides when it's not local (you're providing `namespace/svc-name` instead of K8S DNS `svc-name.namespace` for the URL hostname).
* Using a Service requires a code solution for determining the protocol and appropriate port, which are currently encoded directly in the URL. This would likely require additional constraints on the Service (we'd need a known name scheme for the HTTP and HTTPS admin ports).
* Using a Service is not compatible with the localhost admin API approach, which we want to support for its simple security model.
* Backwards compatibility with the existing URL flag requires that we add code to negotiate between the two options.

IMO we'd only want to add the Service option if we had some functionality that required interactions with the Kubernetes API (e.g. we want to get endpoint information or need to edit its metadata for some reason), and I don't think we have any such functionality proposed. Between that and the other points above, I'd recommend we close this and continue the manifest changes separately in #1500 and other issues attached to #1256.

ccfishk commented:
> #1500



> The assumptions in the various all-in-one YAMLs are by design: those are intentionally opinionated configurations for a select few common configurations (OSS or Enterprise, either with or without a database). They use a single Pod with one Kong container and a controller container because that topology is simple to manage, suitable for many   ##users (especially those that don't have a known reason to use some other topology), and has security benefits for OSS (placing the controller in the same Pod means you don't need to expose the admin API, which is a simple means of securing it without RBAC).

[Jimin] No doubt about all-in-one benefits - simplicity. Stick to design, the Kong Admin TLS option https://github.com/Kong/charts/blob/main/charts/kong/values.yaml#L117 provides the scalability to allow other components  access the service either in cluster or out of the cluster with security. Also, from standardization point of view, each component/service/controller deploy separately provides the topology management flexibility, rather than simplicity.

> 
> If you want another topology, we currently instruct you to use the chart or to build your own manifests by hand. That stems from a combination of historical reasons (when we started providing manifests, Helm was    ###fairly mature and both Kustomize and Operators were new) and capacity (we have limited resources to maintain deployment tools, so have not built out more complex Kustomize manifests or a non-Helm Operator). We don't provide plain manifests for topologies other than single-Pod because there are so many of them (you can basically mix and match any of the ~12 options covered in the [chart documentation](https://github.com/Kong/charts/blob/main/charts/kong/README.md#deployment-options), and even the chart doesn't commit to handling all topologies in their entirety--many require multiple releases).

[Jimin] OK. ATM, most users are using HELM style deployment, users might not be familiar with kubectl tool deployment. I think the kubectl apply -f resource.yaml tool gives user the options to modify resources (namespace etc) on the fly - easy. Yes, we also need consider the chart compitablity. 

> 
> That all said, I still don't think we gain anything by building admin API information from a Service rather than providing a URL, and introduces new problems (which aren't solved here yet):

[Jimin] Either from service, or URL, the point is to avoid hardcode configuration - localhost:8444; 127.0.0.1:8001. Need also use the deployment configuration if there is .
 
> 
> * Using a Service name/namespace doesn't remove the requirement to indicate where the Service resides when it's not local (you're providing `namespace/svc-name` instead of K8S DNS `svc-name.namespace` for the URL hostname).

[Jimin] We can use svc-name.namespace if the Admin is deployed from a cloud provider with valid HostName as external IP. Thats a valid point - both IP and HostName.  Will change accordingly.

> * Using a Service requires a code solution for determining the protocol and appropriate port, which are currently encoded directly in the URL. This would likely require additional constraints on the Service (we'd need a known name scheme for the HTTP and HTTPS admin ports).
[Jimin] Yes, we need the function ConfigService to discover what protocol is supported, and thus use it accordingly.

> * Using a Service is not compatible with the localhost admin API approach, which we want to support for its simple security model.
> * Backwards compatibility with the existing URL flag requires that we add code to negotiate between the two options.
> 
[Jimin] we do provide well-defined compatibility https://github.com/Kong/kubernetes-ingress-controller/pull/1487/files#diff-777c5b378220515f6b692537382429104081d22a761062c7e2f0919e6c0d1757R221
If  in-cluster deployment does not specify any configuration (very likely use the one POD topology), the default localhost will be maintained. 

> IMO we'd only want to add the Service option if we had some functionality that required interactions with the Kubernetes API (e.g. we want to get endpoint information or need to edit its metadata for some reason), and I don't think we have any such functionality proposed. Between that and the other points above, I'd recommend we close this and continue the manifest changes separately in #1500 and other issues attached to #1256.
[Jimin] I think this PR and #1500 provides different solutions to different use cases which direct us to k8s standardization.  I would suggest to de-duplication and merge them accordingly.

ccfishk commented:
Had offline discussion with  @mflendrich, also @rainest's input  - users get used to helm charts for deployment mainly ATM. 
We'd better fix helm charts deploying proxy and controller separately, and use this yaml deployment as a backup for who get used to kubectl tool and our developer folks.

Change the PR to WIP.

ccfishk commented:
> > > unblock controller manager in cluster deployment with appropriate configuration which is illustrated in the yaml and readme file.
> > 
> > 
> > I was trying to understand what issue we had there though--if there is some error that appears, we'd want to understand why it occurs to ensure we're adding a fix. I wasn't aware of any, as when I test the controller 2.0.0-alpha.1 image with the current chart and the all-in-one manifests we have in https://github.com/Kong/kubernetes-ingress-controller/tree/main/deploy/single it does deploy successfully. I am aware of issues with the RBAC config and lack of UDPIngress in the existing examples, but I don't think they cause failures (we just can't update finalizers and have to disable UDPIngress), and we planned to update that elsewhere (#1256 and [Kong/charts#364](https://github.com/Kong/charts/pull/364)). The rest of the manifests that we had for 1.x do still work with 2.x.

[Jimin] @rainest here is the answer to the question "what issue we had there", update 1.3.x all-in-one as a base
https://raw.githubusercontent.com/Kong/kubernetes-ingress-controller/master/deploy/single/all-in-one-dbless.yaml
or using  https://raw.githubusercontent.com/Kong/kubernetes-ingress-controller/e1846684cc23ae5d7530fcb16f4545773908ccc5/deploy/single/all-in-one-postgres.yaml
w/ updating udpingress (CRD, and updating following
- apiGroups:
  - configuration.konghq.com
  resources:
  - udpingresses/status
  verbs:
  - update)

deployment throws following following error messages

time="2021-07-09T01:19:57Z" level=error msg="Reconciler error" error="secrets \"default-token-m5gft\" is forbidden: User \"system:serviceaccount:kong:kong-serviceaccount\" cannot update resource \"secrets\" in API group \"\" in the namespace \"metallb-system\"" logger=controller-runtime.manager.controller.secret name=default-token-m5gft namespace=metallb-system reconciler group= reconciler kind=Secret

> > > The points of the Kong Admin API URL is to remove hardcode ip address (publishStatusAddress) and leverage environment variable retrieving service information. Understand the previous configuration due to controller and proxy are in the same namespace and same pods.
> > 
> > 
> > This sounds related to the work in #1451, correct? [fe24a71](https://github.com/Kong/kubernetes-ingress-controller/commit/fe24a710b41ea07d48688f9290706309d8762ea3) should have already given us what we need there. `publishStatusAddress` should remain: it serves as a user override for Ingress status addresses when they know there network topology doesn't let the controller effectively determine the correct address information (similar to the [Service clusterIP setting](https://kubernetes.io/docs/concepts/services-networking/service/#choosing-your-own-ip-address)), as an alternative to `--publish-service`.
> > That doesn't have any interaction with the admin API, as we're only concerned with the proxy for Ingress status. We shouldn't need any additional admin API logic--while building configuration from a Service instead is viable, it doesn't really give us anything obvious over requesting a URL instead (since we don't need to look up endpoints or obtain other detailed information about the Service), and our documentation all already refers to the existing `--kong-admin-url` flag.
> 
> > > unblock controller manager in cluster deployment with appropriate configuration which is illustrated in the yaml and readme file.
> > 
> > 
> > I was trying to understand what issue we had there though--if there is some error that appears, we'd want to understand why it occurs to ensure we're adding a fix. I wasn't aware of any, as when I test the controller 2.0.0-alpha.1 image with the current chart and the all-in-one manifests we have in https://github.com/Kong/kubernetes-ingress-controller/tree/main/deploy/single it does deploy successfully. I am aware of issues with the RBAC config and lack of UDPIngress in the existing examples, but I don't think they cause failures (we just can't update finalizers and have to disable UDPIngress), and we planned to update that elsewhere (#1256 and [Kong/charts#364](https://github.com/Kong/charts/pull/364)). The rest of the manifests that we had for 1.x do still work with 2.x.
> > > The points of the Kong Admin API URL is to remove hardcode ip address (publishStatusAddress) and leverage environment variable retrieving service information. Understand the previous configuration due to controller and proxy are in the same namespace and same pods.
> > 
> > 
> > This sounds related to the work in #1451, correct? [fe24a71](https://github.com/Kong/kubernetes-ingress-controller/commit/fe24a710b41ea07d48688f9290706309d8762ea3) should have already given us what we need there. `publishStatusAddress` should remain: it serves as a user override for Ingress status addresses when they know there network topology doesn't let the controller effectively determine the correct address information (similar to the [Service clusterIP setting](https://kubernetes.io/docs/concepts/services-networking/service/#choosing-your-own-ip-address)), as an alternative to `--publish-service`.
> > That doesn't have any interaction with the admin API, as we're only concerned with the proxy for Ingress status. We shouldn't need any additional admin API logic--while building configuration from a Service instead is viable, it doesn't really give us anything obvious over requesting a URL instead (since we don't need to look up endpoints or obtain other detailed information about the Service), and our documentation all already refers to the existing `--kong-admin-url` flag.
> 
> Agree that both publish-service and PublicStatusAddress should retain as it is, which was covered by #1509 and #1451. It is unrelated to the KIC individual deployment yaml
> 
> KIC does need KongAdminAPI for configuration communication,
> https://github.com/Kong/kubernetes-ingress-controller/blob/main/railgun/manager/run.go#L96
> 
> Current configuration way https://github.com/Kong/kubernetes-ingress-controller/pull/1500/files#L90 assumes all modules are in the same node and having static port (8444), this will break if we deploy controller and proxy separately. The function retrieve Admin ip and port through namespace and name configuration. https://github.com/Kong/kubernetes-ingress-controller/pull/1487/files#diff-786a1651aa544e8c2f84a32b474758ecb0f8e6c5d9d745b5f643d3d786bc57c4R152

#

rainest commented:
tl;dr the permissions failures are because the controller wanted to add finalizers, and couldn't after I removed resource-wide edit permissions. There are subresource permissions that we could add instead to fix this, but we no longer need them following the removal of finalizers in #1522.

@ccfishk are you still seeing deploy failures with latest next? For testing/pending completion of #1256 1.x manifests should work with:

```
kubectl apply -f https://raw.githubusercontent.com/Kong/kubernetes-ingress-controller/next/railgun/config/crd/bases/configuration.konghq.com_udpingresses.yaml
sed -e "s/name: manager-role/name: kong-ingress-clusterrole/g" railgun/config/rbac/role.yaml | kubectl apply -f -
```
---

These errors:

```
time="2021-07-09T01:19:57Z" level=error msg="Reconciler error" error="secrets "default-token-m5gft" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot update resource "secrets" in API group "" in the namespace "metallb-system"" logger=controller-runtime.manager.controller.secret name=default-token-m5gft namespace=metallb-system reconciler group= reconciler kind=Secret
```
come from adding finalizers to resources, which is a metadata update. We originally had broader permissions on resources, and I reduced permissions on most resources to `get,list,watch` (read-only) earlier to try and match 1.x (for lack of any known reason that we now needed edit permissions also): https://github.com/Kong/kubernetes-ingress-controller/blob/2.0.0-alpha.2/railgun/hack/generators/controllers/networking/main.go#L398-L400

We no longer will need those since we've removed finalizers in https://github.com/Kong/kubernetes-ingress-controller/pull/1522

There's a bit of (possible) inaccuracy in the kubebuilder examples, where https://book.kubebuilder.io/reference/using-finalizers.html indicates you should have full edit permissions on the resource for finalizers, but it looks like [those aren't necessary](https://kubernetes.slack.com/archives/CAR30FCJZ/p1625860261429300). You can handle the finalizer RBAC needs with `update,create,delete` on the `<resource>/finalizer` subresource instead of the parent resource.

If we were to continue using them, we'd want to add that to the kubebuilder RBAC tags in the `railgun/hack/generators/controllers/networking/main.go` template. We wouldn't want to modify the deployment manifests directly, as the next run of the manifest generator (`make generate` followed by `make manifests`) would wipe that out, and wouldn't want to use [the permissions proposed here](https://github.com/Kong/kubernetes-ingress-controller/blob/7973c503f9e756008391edfb91021cc7f5aa1a08/railgun/deployment/kong-ingress-controller/rbac.yaml#L8-L18) (those would grant the controller superuser rights; we want it to only have the permissions it actually needs).

ccfishk commented:
> tl;dr the permissions failures are because the controller wanted to add finalizers, and couldn't after I removed resource-wide edit permissions. There are subresource permissions that we could add instead to fix this, but we no longer need them following the removal of finalizers in #1522.
> 
> @ccfishk are you still seeing deploy failures with latest next? For testing/pending completion of #1256 1.x manifests should work with:
> 
> ```
> kubectl apply -f https://raw.githubusercontent.com/Kong/kubernetes-ingress-controller/next/railgun/config/crd/bases/configuration.konghq.com_udpingresses.yaml
> sed -e "s/name: manager-role/name: kong-ingress-clusterrole/g" railgun/config/rbac/role.yaml | kubectl apply -f -
> ```


> 
> These errors:
> 
> ```
> time="2021-07-09T01:19:57Z" level=error msg="Reconciler error" error="secrets "default-token-m5gft" is forbidden: User "system:serviceaccount:kong:kong-serviceaccount" cannot update resource "secrets" in API group "" in the namespace "metallb-system"" logger=controller-runtime.manager.controller.secret name=default-token-m5gft namespace=metallb-system reconciler group= reconciler kind=Secret
> ```
> 
> come from adding finalizers to resources, which is a metadata update. We originally had broader permissions on resources, and I reduced permissions on most resources to `get,list,watch` (read-only) earlier to try and match 1.x (for lack of any known reason that we now needed edit permissions also): https://github.com/Kong/kubernetes-ingress-controller/blob/2.0.0-alpha.2/railgun/hack/generators/controllers/networking/main.go#L398-L400

  [Jimin] AFAIK, finalizer only applied for resource deletion, we apply finalizer for KIC ingress we created. If your operation include all namespaces/any resources, might be associated to this error.  Not sure why only got this complain at that moment, would like to take one more look/dig into my case. 
> 
> We no longer will need those since we've removed finalizers in #1522
> 
> There's a bit of (possible) inaccuracy in the kubebuilder examples, where https://book.kubebuilder.io/reference/using-finalizers.html indicates you should have full edit permissions on the resource for finalizers, but it looks like [those aren't necessary](https://kubernetes.slack.com/archives/CAR30FCJZ/p1625860261429300). You can handle the finalizer RBAC needs with `update,create,delete` on the `<resource>/finalizer` subresource instead of the parent resource.
> 
> If we were to continue using them, we'd want to add that to the kubebuilder RBAC tags in the `railgun/hack/generators/controllers/networking/main.go` template. We wouldn't want to modify the deployment manifests directly, as the next run of the manifest generator (`make generate` followed by `make manifests`) would wipe that out, and wouldn't want to use [the permissions proposed here](https://github.com/Kong/kubernetes-ingress-controller/blob/7973c503f9e756008391edfb91021cc7f5aa1a08/railgun/deployment/kong-ingress-controller/rbac.yaml#L8-L18) (those would grant the controller superuser rights; we want it to only have the permissions it actually needs).
[Jimin] aware of the permission scope - it is not Kong's cluster, user's cluster. Agree restrict permission is good. thansk.
#

> https://github.com/Kong/kubernetes-ingress-controller/blob/2.0.0-alpha.2/railgun/hack/generators/controllers/networking/main.go#L398-L400

rainest commented:
> Not sure why only got this complain at that moment, would like to take one more look/dig into my case.

Some of the errors from controller-gen reconciler code are unfortunately rather opaque--they tell you they're trying to edit an object and failed, but not the specific edit and/or where in the code it logged the error.

Don't actually recall how I realized the finalizers were the culprit last week. I observed that adding

```
- apiGroups:
  - whatever
  resources:
  - whatever/finalizers
  verbs:
  - update
  - create
  - delete
```
to roles cleared it, but forget why I thought to do that--I think it was [reading through the generated reconciler](https://github.com/Kong/kubernetes-ingress-controller/blob/cb27c4b2d83abb068c55d03a16f0602a77f89cae/railgun/controllers/configuration/zz_generated_controllers.go#L212-L258) and seeing which successful logs appeared and checking what client actions it took after. Doing that is rather tedious.

ccfishk commented:
Close this atm according to discussion last Friday.
